File: Graphs/Detect cycle in an undirected graph/ans2.cpp
class Solution {
  public:
    bool iscycle(vector<int> adj[], vector<bool> &visited, int i, int parent){
    queue<pair<int, int>> pq;
    visited[i] = true;
    pq.push({i, -1});
    while(!pq.empty())
    {
        pair<int, int> p = pq.front();
        int a = p.first;
        int b = p.second;
        pq.pop();
        for(auto it : adj[a])
        {
            if(it != b)
            {
                if(visited[it])
                return true;
                else
                {
                    visited[it] = true;
                    pq.push({it, a});
                }
            }
        }
    }
    return false;
    }
    bool isCycle(int n, vector<int> adj[]) {
    
        vector<bool> visited(n, 0);
        for(int i=0; i<n; i++)
        {
            if(!visited[i] && iscycle(adj, visited, i, -1))
            return true;
        }
        return false;
    }
};

File: Graphs/Is Graph Bipartite/ans2.py
class Solution(object):
    def isBipartite(self, graph):
        n = len(graph)
        color = [-1]*n

        for i in range(n):
            if color[i] == -1:
                color[i] = 1
                queue = [i]
                while len(queue) > 0:
                    t = queue.pop(0)
                    for x in graph[t]:
                        if color[x] == -1:
                            color[x] = 1 - color[t]
                            queue.append(x)
                        elif color[x] == color[t]:
                            return False
        
        return True